Add uninstall cmd functionality to cli (#3622)

Matei207 opened the pull request:
### Fixes #3622 
### Edit 29th March: 
PR overhauled, updated body of this msg accordingly

Signed-off-by: Matei David <matei.david.35@gmail.com>

### What
---

* Introduce an `uninstall` command to the cli to generate resources necessary to purge/uninstall Linkerd from a Kubernetes cluster (uninstall was chosen over delete for the sake of brevity and conciseness in commands);

### How
---

* `uninstall` uses `client-go` to fetch all `cluster-wide resources` (RBAC, hooks, etc.) and the control-plane `namespace` and outputs them to stdout as yaml manifests (to be piped to `kubectl delete`
* clientset is used to grab all resources that have the `linkerd.io/control-plane-ns` label. For the `namespace` itself, the name is used instead of a label selector.
* Since `gvk` is stripped down when fetching a Kubernetes resource using `client-go` ([see here](https://github.com/kubernetes/client-go/issues/541#issuecomment-452312901)), the `marshal` function also infers the `GVK` from the object type in case it is missing.

### Test
* Test case is quite trivial, however, it ensures that empty files (as strings) are not generated, and it ensures the `apiVersion` and `kind` are correctly injected/present (without them, the pipe to `kubectl` will fail).

* I checked the results manually by following the steps below (see attached pictures):

1. Perform a fresh install through `linkerd install | kubectl apply -f -`
2. Run  `kubectl get APIService,ClusterRoleBinding,CustomResourceDefinition,MutatingWebhookConfiguration,ValidatingWebhookConfiguration,PodSecurityPolicy,Namespace  -l linkerd.io/control-plane-ns=linkerd -o yaml` -- this outputs all resources that should be deleted (except `RoleBIndings` in `kube-system` namespace which have been added as part of this PR)
3. Run `/bin/go-run cli uninstall | kubectl delete -f - `
4. Repeat step (2) -- output is now empty list

---
![Screen Shot 2020-03-29 at 4 02 31 PM](https://user-images.githubusercontent.com/18265670/77852547-2459e580-71d7-11ea-951b-9a259868c68a.png)
![Screen Shot 2020-03-29 at 4 05 22 PM](https://user-images.githubusercontent.com/18265670/77852550-26bc3f80-71d7-11ea-9726-a005d5a01297.png)



<!--  Thanks for sending a pull request!

If you already have a well-structured git commit message, chances are GitHub
set the title and description of this PR to the git commit message subject and
body, respectively. If so, you may delete these instructions and submit your PR.

If this is your first time, please read our contributor guide:
https://github.com/linkerd/linkerd2/blob/master/CONTRIBUTING.md

The title and description of your Pull Request should match the git commit
subject and body, respectively. Git commit messages are structured as follows:

```
Subject

Problem

Solution

Validation

Fixes #[GitHub issue ID]

DCO Sign off
```

Example git commit message:

```
Introduce Pull Request Template

GitHub's community guidelines recommend a pull request template, the repo was
lacking one.

Introduce a `PULL_REQUEST_TEMPLATE.md` file.

Once merged, the
[Community profile checklist](https://github.com/linkerd/linkerd2/community)
should indicate the repo now provides a pull request template.

Fixes #3321

Signed-off-by: Jane Smith <jane.smith@example.com>
```

Note the git commit message subject becomes the pull request title.

For more details around git commits, see the section on Committing in our
contributor guide:
https://github.com/linkerd/linkerd2/blob/master/CONTRIBUTING.md#committing
-->

alpeb commented:
Thank you @Matei207 for all your effort. I'm gonna add some comments into the original issue #3622.

adleong commented on cli/cmd/uninstall.go:
```diff
@@ -0,0 +1,304 @@
+	return nonEmptyResources
+}
+
+func fetchClusterRoles(k *k8s.KubernetesAPI, options metav1.ListOptions) (string, error) {
```
I notice that we have a mix of list and non-list type resources in the output, which probably works, but is a bit confusing.  I'd prefer to just have a sequence of non-list resources.  I think we can accomplish that by having all of these `fetch...` functions return a slice of `ObjectReference` or something like that.

Matei207 replied:
@adleong great suggestion will get it underway! Since we only need kinds, name and version, I thought maybe it's better to have our own wrapper objects around the metadata we want to marshal? In that case, all functions can return a slice of "wrapperStruct".

adleong commented on cli/cmd/uninstall.go:
```diff
@@ -0,0 +1,304 @@
+
+func marshalKubernetesResource(resource runtime.Object) (string, error) {
+	var gvk = resource.GetObjectKind().GroupVersionKind()
+	if gvk.Empty() {
```
I'm not sure what's going on here with populating the gvk.  Can you explain this?

Matei207 replied:
@adleong of course! I realise how unnecessarily complex this looks.

Using `client-go`, whenever we fetch a resource, the `TypeMeta` seems to be stripped from it. I haven't managed to find a reason why this happens, but going through some related issues, it seems to be an intentional design; [#541 in client-go](https://github.com/kubernetes/client-go/issues/541) is an example of an issue where this problem is brought up to light.

I found a helper function that returns the gvk by using reflection on the object type, and that's the solution I went for. Prior to this, I tried to inject the gvk myself, this seemed to be a bit cleaner, however.

If there's another solution for it, I'll get it changed asap.

adleong commented on cli/cmd/uninstall.go:
```diff
@@ -0,0 +1,304 @@
+		}
+		resource.GetObjectKind().SetGroupVersionKind(gvk)
+	}
+	b, err := yaml.Marshal(resource)
```
Technically, we only need to marshal the name, namespace (if applicable), kind, and apiVersion, right?  I think we can maybe get away with just marshalling an ObjectReference instead of the entire resource.

Matei207 replied:
@adleong yes, completely right! It was an oversight on my part, was under the impression it did not work just with the kind, name and apiVersion. Will make changes.

Matei207 commented:
I have updated the PR with the above comments @adleong.

I have added two new structs to reduce the verbosity when marshaling. As you suggested, all of the fetch... functions (except the namespace one) now return a slice of `kubernetesResource`.

The resource itself holds the `apiVersion`, the `kind` and the `namespace` and `name` (both under a different metadata struct). The only one that needs to use the namespace is the `RoleBinding` object.

There is a bit of repetitiveness going around that I have no idea how to solve (particularly around turning list objects into kubernetesResource types), even if I would have that as a separate method that accepts an interface{}, all list objects would have to be separately cast.

I hope this is satisfactory, if not, keen on seeing what else can be improved!

Matei207 commented on cli/cmd/uninstall.go:
```diff
@@ -0,0 +1,399 @@
+	cmd := &cobra.Command{
+		Use:   "uninstall",
+		Args:  cobra.NoArgs,
+		Short: "Output Kubernetes resources to uninstall Linkerd",
```
Done, thank you.

Matei207 commented:
Thank you for your comment, @alpeb .

I have now added an integration test like you have suggested. It occurs before `TestInstall_sp` and compares the output from `uninstall` with the golden file.

The integration tests (`custom_domain`) kept failing since before I have introduced anything new to them, not sure if this is just undefined behaviour or something I have introduced. I have rebased off master and pushed again to get rid of all of the excess commits and re-trigger the build.

If there is anything else you'd like changed, let me know.

alpeb commented:
The integration test failure was introduced by #4212. I'm working on a fix.

adleong commented on cli/cmd/uninstall.go:
```diff
@@ -0,0 +1,399 @@
+		Metadata   kubernetesMetadata `json:"metadata"`
+	}
+
+	kubernetesMetadata struct {
```
can we use [ObjectMeta](https://godoc.org/k8s.io/apimachinery/pkg/apis/meta/v1#ObjectMeta) instead?

Matei207 replied:
The embedded fields made everything much better, amazing suggestion, thank you.

adleong commented on cli/cmd/uninstall.go:
```diff
@@ -0,0 +1,399 @@
+
+type (
+	kubernetesResource struct {
+		APIVersion string             `json:"apiVersion"`
```
Can these fields be an embedded TypeMeta struct instead?

adleong commented on cli/cmd/uninstall.go:
```diff
@@ -0,0 +1,399 @@
+
+	resources := make([]kubernetesResource, len(list.Items))
+	for i, item := range list.Items {
+		v, k := getObjectGvk(item.DeepCopy()).ToAPIVersionAndKind()
```
It's annoying that the kubernetes client strips out the gvk like this.  However, we actually statically know the gvk already since we know we're dealing with cluster roles here.  We can manually set the kind and apiVersion to `ClusterRole` and `rbac.authorization.k8s.io/v1`.  And similarly for each other resource type, we know their kind and apiVersion ahead of time so we can manually fill in those fields.

Matei207 replied:
Done. It reads much better now, thank you.

adleong commented on cli/cmd/uninstall.go:
```diff
@@ -0,0 +1,399 @@
+		return nil, err
+	}
+
+	if list.Items == nil {
```
I believe that if a slice is nil, the for-loop will skip right over it, so I don't think we need this special case.

adleong commented on cli/cmd/uninstall.go:
```diff
@@ -0,0 +1,288 @@
+}
+
+func newKubernetesResource(apiVersion, kind, name string) kubernetesResource {
+	r := kubernetesResource{}
```
I think we can inline this struct initialization:

```
return kubernetesResource {
  runtime.TypeMeta {
    APIVersion: apiVersion,
    Kind: kind,
  },
  metav1.ObjectMeta {
    Name: name,
  },
}
```

Matei207 replied:
Changed. Good to know how embedded types behave in go, thought (for some reason) that wouldn't work quite well, but turns out it does 👍 .

Matei207 commented on cli/cmd/uninstall.go:
```diff
@@ -0,0 +1,288 @@
+
+const (
+	yamlSep                   = "---\n"
+	rbacVersion1              = "rbac.authorization.k8s.io/v1"
```
Took the suggestion into account.
